File: sdks/php/src/Lexer.php
<?php

declare(strict_types=1);

namespace MikeMackintosh\CrowdControl;

/**
 * Ports github.com/mikemackintosh/crowdcontrol/parser/lexer.go to PHP.
 */

enum TokenType: int
{
    case Eof = 0;
    case Ident = 1;
    case StringT = 2;
    case Number = 3;
    case LBrace = 4;
    case RBrace = 5;
    case LBracket = 6;
    case RBracket = 7;
    case LParen = 8;
    case RParen = 9;
    case Dot = 10;
    case Comma = 11;
    case Eq = 12;
    case Neq = 13;
    case Lt = 14;
    case Gt = 15;
    case Lte = 16;
    case Gte = 17;
    case Star = 18;
    case Plus = 19;
    case Minus = 20;
    case Slash = 21;
}

final class Token
{
    public function __construct(
        public TokenType $type,
        public string $val = '',
        public int $line = 0,
        public int $col = 0,
    ) {
    }

    public function __toString(): string
    {
        return $this->type === TokenType::Eof ? 'EOF' : "'" . $this->val . "'";
    }
}

final class LexException extends \RuntimeException
{
}

final class Lexer
{
    /** @var array<int, string> */
    private array $chars;
    private int $pos = 0;
    private int $line = 1;
    private int $col = 1;
    /** @var list<Token> */
    private array $tokens = [];

    public function __construct(string $source)
    {
        // Split into Unicode code points so multi-byte characters work.
        $this->chars = preg_split('//u', $source, -1, PREG_SPLIT_NO_EMPTY) ?: [];
    }

    /**
     * @return list<Token>
     */
    public static function tokenize(string $source): array
    {
        return (new self($source))->run();
    }

    /**
     * @return list<Token>
     */
    public function run(): array
    {
        $n = count($this->chars);
        while ($this->pos < $n) {
            $ch = $this->chars[$this->pos];

            // Whitespace
            if ($ch === ' ' || $ch === "\t" || $ch === "\r" || $ch === "\n") {
                if ($ch === "\n") {
                    $this->line++;
                    $this->col = 1;
                } else {
                    $this->col++;
                }
                $this->pos++;
                continue;
            }

            // Comments: # or //
            if ($ch === '#' || ($ch === '/' && $this->pos + 1 < $n && $this->chars[$this->pos + 1] === '/')) {
                while ($this->pos < $n && $this->chars[$this->pos] !== "\n") {
                    $this->pos++;
                }
                continue;
            }

            // String literal
            if ($ch === '"') {
                $this->tokens[] = $this->lexString();
                continue;
            }

            // Number
            if (ctype_digit($ch)) {
                $this->tokens[] = $this->lexNumber();
                continue;
            }

            // Identifier / keyword
            if (ctype_alpha($ch) || $ch === '_') {
                $this->tokens[] = $this->lexIdent();
                continue;
            }

            // Two-char operators
            if ($this->pos + 1 < $n) {
                $two = $ch . $this->chars[$this->pos + 1];
                $mapping = [
                    '==' => TokenType::Eq,
                    '!=' => TokenType::Neq,
                    '<=' => TokenType::Lte,
                    '>=' => TokenType::Gte,
                ];
                if (isset($mapping[$two])) {
                    $this->tokens[] = new Token($mapping[$two], $two, $this->line, $this->col);
                    $this->pos += 2;
                    $this->col += 2;
                    continue;
                }
            }

            // Single-char tokens
            $single = [
                '{' => TokenType::LBrace,
                '}' => TokenType::RBrace,
                '[' => TokenType::LBracket,
                ']' => TokenType::RBracket,
                '(' => TokenType::LParen,
                ')' => TokenType::RParen,
                '.' => TokenType::Dot,
                ',' => TokenType::Comma,
                '<' => TokenType::Lt,
                '>' => TokenType::Gt,
                '*' => TokenType::Star,
                '+' => TokenType::Plus,
                '-' => TokenType::Minus,
                '/' => TokenType::Slash,
            ];
            if (isset($single[$ch])) {
                $this->tokens[] = new Token($single[$ch], $ch, $this->line, $this->col);
                $this->pos++;
                $this->col++;
                continue;
            }

            throw new LexException("line {$this->line} col {$this->col}: unexpected character: {$ch}");
        }

        $this->tokens[] = new Token(TokenType::Eof, '', $this->line, $this->col);
        return $this->tokens;
    }

    private function lexString(): Token
    {
        $startCol = $this->col;
        $this->pos++; // skip opening "
        $this->col++;
        $parts = '';
        $n = count($this->chars);
        while ($this->pos < $n) {
            $ch = $this->chars[$this->pos];
            if ($ch === '"') {
                $this->pos++;
                $this->col++;
                return new Token(TokenType::StringT, $parts, $this->line, $startCol);
            }
            if ($ch === '\\' && $this->pos + 1 < $n) {
                $this->pos++;
                $this->col++;
                $next = $this->chars[$this->pos];
                $parts .= match ($next) {
                    '"', '\\' => $next,
                    'n' => "\n",
                    't' => "\t",
                    default => '\\' . $next,
                };
            } else {
                $parts .= $ch;
            }
            $this->pos++;
            $this->col++;
        }
        throw new LexException("line {$this->line} col {$startCol}: unterminated string");
    }

    private function lexNumber(): Token
    {
        $start = $this->pos;
        $startCol = $this->col;
        $n = count($this->chars);
        while ($this->pos < $n && (ctype_digit($this->chars[$this->pos]) || $this->chars[$this->pos] === '.')) {
            $this->pos++;
            $this->col++;
        }
        $val = implode('', array_slice($this->chars, $start, $this->pos - $start));
        return new Token(TokenType::Number, $val, $this->line, $startCol);
    }

    private function lexIdent(): Token
    {
        $start = $this->pos;
        $startCol = $this->col;
        $n = count($this->chars);
        while ($this->pos < $n) {
            $ch = $this->chars[$this->pos];
            if (ctype_alnum($ch) || $ch === '_') {
                $this->pos++;
                $this->col++;
            } else {
                break;
            }
        }
        $val = implode('', array_slice($this->chars, $start, $this->pos - $start));
        return new Token(TokenType::Ident, $val, $this->line, $startCol);
    }
}

File: parser/lexer.go
package parser

import (
	"fmt"
	"strings"
	"unicode"
)

// TokenType identifies token kinds.
type TokenType int

const (
	TokenEOF      TokenType = iota
	TokenIdent              // keywords and identifiers: forbid, warn, permit, resource, etc.
	TokenString             // "quoted string"
	TokenNumber             // 123, 5.0
	TokenLBrace             // {
	TokenRBrace             // }
	TokenLBracket           // [
	TokenRBracket           // ]
	TokenLParen             // (
	TokenRParen             // )
	TokenDot                // .
	TokenComma              // ,
	TokenEq                 // ==
	TokenNeq                // !=
	TokenLt                 // <
	TokenGt                 // >
	TokenLte                // <=
	TokenGte                // >=
	TokenStar               // * (also used as multiply in arithmetic)
	TokenPlus               // +
	TokenMinus              // -
	TokenSlash              // /
)

// Token is a lexed token with position info.
type Token struct {
	Type TokenType
	Val  string
	Line int
	Col  int
}

func (t Token) String() string {
	if t.Type == TokenEOF {
		return "EOF"
	}
	return fmt.Sprintf("%q", t.Val)
}

// Lexer tokenizes CrowdControl policy source code.
type Lexer struct {
	input  []rune
	pos    int
	line   int
	col    int
	tokens []Token
}

// Lex tokenizes the input string.
func Lex(input string) ([]Token, error) {
	l := &Lexer{
		input: []rune(input),
		line:  1,
		col:   1,
	}
	if err := l.run(); err != nil {
		return nil, err
	}
	l.tokens = append(l.tokens, Token{Type: TokenEOF, Line: l.line, Col: l.col})
	return l.tokens, nil
}

func (l *Lexer) run() error {
	for l.pos < len(l.input) {
		ch := l.input[l.pos]

		// Skip whitespace
		if unicode.IsSpace(ch) {
			if ch == '\n' {
				l.line++
				l.col = 1
			} else {
				l.col++
			}
			l.pos++
			continue
		}

		// Skip comments
		if ch == '#' || (ch == '/' && l.pos+1 < len(l.input) && l.input[l.pos+1] == '/') {
			for l.pos < len(l.input) && l.input[l.pos] != '\n' {
				l.pos++
			}
			continue
		}

		// String literal
		if ch == '"' {
			tok, err := l.lexString()
			if err != nil {
				return err
			}
			l.tokens = append(l.tokens, tok)
			continue
		}

		// Number
		if unicode.IsDigit(ch) {
			l.tokens = append(l.tokens, l.lexNumber())
			continue
		}

		// Identifier or keyword
		if unicode.IsLetter(ch) || ch == '_' {
			l.tokens = append(l.tokens, l.lexIdent())
			continue
		}

		// Two-character operators
		if l.pos+1 < len(l.input) {
			two := string(l.input[l.pos : l.pos+2])
			switch two {
			case "==":
				l.tokens = append(l.tokens, Token{Type: TokenEq, Val: two, Line: l.line, Col: l.col})
				l.pos += 2
				l.col += 2
				continue
			case "!=":
				l.tokens = append(l.tokens, Token{Type: TokenNeq, Val: two, Line: l.line, Col: l.col})
				l.pos += 2
				l.col += 2
				continue
			case "<=":
				l.tokens = append(l.tokens, Token{Type: TokenLte, Val: two, Line: l.line, Col: l.col})
				l.pos += 2
				l.col += 2
				continue
			case ">=":
				l.tokens = append(l.tokens, Token{Type: TokenGte, Val: two, Line: l.line, Col: l.col})
				l.pos += 2
				l.col += 2
				continue
			}
		}

		// Single character tokens
		tok := Token{Line: l.line, Col: l.col, Val: string(ch)}
		switch ch {
		case '{':
			tok.Type = TokenLBrace
		case '}':
			tok.Type = TokenRBrace
		case '[':
			tok.Type = TokenLBracket
		case ']':
			tok.Type = TokenRBracket
		case '(':
			tok.Type = TokenLParen
		case ')':
			tok.Type = TokenRParen
		case '.':
			tok.Type = TokenDot
		case ',':
			tok.Type = TokenComma
		case '<':
			tok.Type = TokenLt
		case '>':
			tok.Type = TokenGt
		case '*':
			tok.Type = TokenStar
		case '+':
			tok.Type = TokenPlus
		case '-':
			tok.Type = TokenMinus
		case '/':
			// Check it's not a comment (// already handled above)
			tok.Type = TokenSlash
		default:
			return fmt.Errorf("line %d col %d: unexpected character: %c", l.line, l.col, ch)
		}
		l.tokens = append(l.tokens, tok)
		l.pos++
		l.col++
	}

	return nil
}

func (l *Lexer) lexString() (Token, error) {
	start := l.pos
	startCol := l.col
	l.pos++ // skip opening "
	l.col++

	var sb strings.Builder
	for l.pos < len(l.input) {
		ch := l.input[l.pos]
		if ch == '"' {
			l.pos++
			l.col++
			return Token{Type: TokenString, Val: sb.String(), Line: l.line, Col: startCol}, nil
		}
		if ch == '\\' && l.pos+1 < len(l.input) {
			l.pos++
			l.col++
			next := l.input[l.pos]
			switch next {
			case '"', '\\':
				sb.WriteRune(next)
			case 'n':
				sb.WriteRune('\n')
			case 't':
				sb.WriteRune('\t')
			default:
				sb.WriteRune('\\')
				sb.WriteRune(next)
			}
		} else {
			sb.WriteRune(ch)
		}
		l.pos++
		l.col++
	}

	return Token{}, fmt.Errorf("line %d col %d: unterminated string", l.line, start)
}

func (l *Lexer) lexNumber() Token {
	start := l.pos
	startCol := l.col
	for l.pos < len(l.input) && (unicode.IsDigit(l.input[l.pos]) || l.input[l.pos] == '.') {
		l.pos++
		l.col++
	}
	return Token{Type: TokenNumber, Val: string(l.input[start:l.pos]), Line: l.line, Col: startCol}
}

func (l *Lexer) lexIdent() Token {
	start := l.pos
	startCol := l.col
	for l.pos < len(l.input) && (unicode.IsLetter(l.input[l.pos]) || unicode.IsDigit(l.input[l.pos]) || l.input[l.pos] == '_') {
		l.pos++
		l.col++
	}
	return Token{Type: TokenIdent, Val: string(l.input[start:l.pos]), Line: l.line, Col: startCol}
}
